### PR TITLE
Laboratory work #2, Kurnakova Irina - 20FPL1

### PR DESCRIPTION
**Pull request by IrinaKurnakova**

### CODE REVIEW

File: lab_2/main.py [hunk 1]
```diff
@@ -13,7 +13,20 @@ def get_freq_dict(tokens: list) -> dict or None:
...
     :return: a dictionary with frequencies
     """
-    pass
+    if not isinstance(tokens, list):
+        return None
+    for i in tokens:
+        if not isinstance(i, str):
+            return None
+    freq_dict = {}
+    for i in tokens:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:22]: better to use names like token/word/etc

File: lab_2/main.py [hunk 2]
```diff
@@ -24,16 +37,34 @@ def get_language_profiles(texts_corpus: list, language_labels: list) -> dict or
...
+        if not isinstance(corpus, list):
+            return None
+        for token in corpus:
+            if not isinstance(token, str):
+                return None
+    for label in language_labels:
+        if not isinstance(label, str):
+            return None
+    language_profiles = {key: get_freq_dict(value) for key, value
+                         in zip(language_labels, texts_corpus)}
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:52]: what happens on this line?

File: lab_2/main.py [hunk 3]
```diff
@@ -43,8 +74,21 @@ def get_text_vector(original_text: list, language_profiles: dict) -> list or Non
...
-    pass
-
+    if not isinstance(original_text, list) or not isinstance(language_profiles, dict):
+        return None
+    text_vector = []
+    for word in get_language_features(language_profiles):
+        if word in original_text:
+            for freq_dict in language_profiles.values():
+                for key in freq_dict:
+                    if key == word:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:84]: for for for, a bit hard to read, any other suggestions?

File: lab_2/main.py [hunk 4]
```diff
@@ -64,8 +121,27 @@ def predict_language_score(unknown_text_vector: list, known_text_vectors: list,
...
+        if not isinstance(i, (float, int)):
+            return None
+    for i in known_text_vectors:
+        if not isinstance(i, list):
+            return None
+    for i in language_labels:
+        if not isinstance(i, str):
+            return None
+    dictionary = {}
+    for i in range(len(known_text_vectors)):
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:139]: same about naming

File: lab_2/main.py [hunk 5]
```diff
@@ -64,8 +121,27 @@ def predict_language_score(unknown_text_vector: list, known_text_vectors: list,
...
+        dictionary[language_labels[i]] = calculate_distance(unknown_text_vector, known_text_vectors[i])
+    min_value = min(dictionary.values())
+    for key, value in dictionary.items():
+        if value == min_value:
+            return [key, value]
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:144]: is there a shorter way to find min?

**Review by ValeriyaKuznetsova — COMMENTED** [review]
looks good, keep working!

File: lab_2/main.py (hunk 2 above)
Comment (IrinaKurnakova) [reply lab_2/main.py:52]: Function zip creates a list of tuples where every tuple is (n-th element of language_labels, n-th element of texts_corpus). With the help of dictionary generator we create a dictionary where the certain key is the first element of the n-th tuple and his value is a frequency dictionary we get using get_freq_dict function (in this case, this function takes the second element of the n-th tuple as an argument).

File: lab_2/main.py [hunk 6]
```diff
@@ -186,7 +239,27 @@ def calculate_distance_sparse(unknown_text_vector: list,
...
     :param known_text_vector: sparse vector for known text
     """
-    pass
+    if not isinstance(unknown_text_vector, list)\
+            or not isinstance(known_text_vector, list):
+        return None
+    for element in unknown_text_vector:
+        if not isinstance(element, list):
+            return None
+    for element in known_text_vector:
```
[the remark this answers is not in the document]
Comment (IrinaKurnakova) [reply lab_2/main.py:248]: The task doesn't say what metric we should use (euclid or manhattan), so I decided to use euclid distance. All the tests passed, so it turned out that it was right decision here.

File: lab_2/start.py [hunk 7]
```diff
@@ -37,7 +39,35 @@
...
+    k = 3
+    RESULT = []
+
+    for text in UNKNOWN_SAMPLES:
+        unknown_text = remove_stop_words(tokenize(text), stop_words)
+        unknown_text_vector = get_sparse_vector(unknown_text, language_profiles)
+        predicted_language = predict_language_knn_sparse(unknown_text_vector, known_text_vectors,
+                                                         language_labels, k)
+        RESULT.append(predicted_language[0])
+
```
[the remark this answers is not in the document]
Comment (IrinaKurnakova) [reply lab_2/start.py:70]: Firstly, we create the profiles of known languages. For this, we handle texts, tokenizing and removing stop words, and we put these tokens into the list texts_corpus. This texts_corpus we use in function get_language_profiles and get the profiles we need. After that, we create known text vectors, usung function get_sparce_vector. Then we work with unknown samples in the same way (tokenize, remove stop words, get vector). Finally, with the help of function predict_language_knn_sparse we get the predicted language (this function returns a list consisting of language label and distance, so we take only label - predicted_language[0]).

**Pull request closed without merging**